core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala

```diff
@@ -38,65 +37,48 @@ private[zio] final class FiberContext[E, A](
   private[this] val state = new AtomicReference[FiberState[E, A]](FiberState.Initial[E, A])
 
   // Accessed from within a single thread (not necessarily the same):
-  @volatile private[this] var noInterrupt = 0
   @volatile private[this] var supervised  = List.empty[Set[FiberContext[_, _]]]
   @volatile private[this] var supervising = 0
   @volatile private[this] var locked      = List.empty[Executor]
   @volatile private[this] var environment = List[Any](())
 
-  private[this] val fiberId = FiberContext.fiberCounter.getAndIncrement()
-  private[this] val stack   = new Stack[Any => IO[Any, Any]]()
+  private[this] val interruptStatus = StackBool()
+  private[this] val fiberId         = FiberContext.fiberCounter.getAndIncrement()
+  private[this] val stack           = new Stack[Any => IO[Any, Any]]()
 
   final def runAsync(k: Callback[E, A]): Unit =
     register0(xx => k(Exit.flatten(xx))) match {
       case null =>
       case v    => k(v)
     }
 
-  private class Finalizer(val finalizer: UIO[_]) extends Function[Any, IO[E, Any]] {
+  private object InterruptExit extends Function[Any, IO[E, Any]] {
     final def apply(v: Any): IO[E, Any] = {
-      noInterrupt += 1
+      interruptStatus.popDrop(())
 
-      finalizer.flatMap(_ => IO.effectTotal { noInterrupt -= 1; v })
+      ZIO.succeed(v)
```

Comment:
I'm pretty sure it would create an extra interruption point in, for example,
```scala
def dequeue = (... foo.interruptible ...).uninterruptible
(dequeue /* HERE */).interruptible.flatMap(process).uninterruptible
```
which can lead to data loss.

Reply:
So, if we are currently `interruptible`, then we should use `pop`+`succeed`, and if not - `effectTotal { pop; v }`

Reply:
As far as I can see, they have the same interruptibility characteristics. Because even with `effectTotal { pop; v }`, that leads to `Effect` being evaluated, which creates a new `curIo`, which then goes through the `shouldInterrupt` loop again before being processed.

This is indeed why it's not safe to put `interruptible` on acquisition (I know I sound like a broken record 😆). You will sometimes lose data. Indeed, you can see even if such things were not interruptible "at the boundary", if you refactored it to `(queue.take <* ZIO.unit).interruptible`, you can see the "harmless" refactoring has now postponed the acquisition by another `FlatMap`, giving time for interruption to occur.

Basically using `interruptible` in this way is not safe, because there's a possibility the acquire succeeds but is interrupted before it returns.

Reply:
Not the same.

In case of `pop` + `succeed`: 
1.`nextInstr` calls `InterruptExit.apply`, which pops `false` from the stack, enabling interruption and returns `ZIO.Succeed`
2. Which immediately gets interrupted in the main loop.

In case of `effectTotal { pop; v }`:
1. `nextInstr` returns `ZIO.Effect`
2. Which is not interrupted, so the main loop enters `ZIO.Effect` case
3. Pops `false` from stack, enabling interruption, and calls `nextInstr`
4. Which, in case of `uninterruptible` followed by `interruptible`, calls `InterruptExit.apply`
5. Which pops `true` from the stack, disabling interruption, and returns `ZIO.Succeed`. The computation continues safely to `flatMap(process)`.

`queue.take <* ZIO.unit` is not "harmless" refactoring vs interruption. ZIO has a great chance to have a clean and well-specified interruptibility behavior (and it is not that hard to implement), with fine-grain control on the combinator level. Haskell, for example, cannot have it, because async exceptions are delivered on memory allocations, not just in-between statements.

Reply:
You are right on the distinction. Basically the atomic return + exit can allow the value to be fed to the continuation on the stack before interruption. Although there is no guarantee that actually changes the situation.

This is not safe to do if the region is interruptible already, but you identified this already:

> So, if we are currently interruptible, then we should use pop+succeed, and if not - effectTotal { pop; v }

I will make the change, but I think maybe it's too weak. I think what we really want to say (at a higher-level) is that coming out of `foo.interruptible`, we're going to guarantee that if `foo` produced a value, then it will be fed to the continuation rather than "dropped at the edge".

That does not solve all problems, see `queue.take <* ZIO.unit`. And while you can say, "that's obviously wrong", what if you peer into the implementation of `queue.take` and see something like:

```scala
def take = 
  for {
    ...
  } yield value
```

This desugars into a final `map` that has the same problem as `queue.take <* ZIO.unit`. Basically this would be sane only in case that the final instruction is an atomic one (e.g. `effect`) that actually does the acquisition. Any seemingly benign refactoring from this state would lead to loss of the resource being acquired in some circumstances.

That's super fragile. I think it's maybe OK for power users but cannot be recommended as a default way of solving "interruptible acquisition".

Reply:
Fixed, but I did not take on "guarantee pass to continuation".

A remaining question is whether `interruptible(uninterruptible(foo)) === uninterruptible(foo)`, and `uninterruptible(interruptible(foo)) === interruptible(foo)`.

This might be a nice guarantee to have. It would mean if you pass in `foo` and make it `interruptible` but it was already `uninterruptible`, then it has no effect.

Reply:
Thanks. Yes, I wrote about the "no extra interruption point" guarantee back in the #580 (`Succeed` and `Effect` are fine, but the `EffectAsync` is not, though it can be fixed). I think that synchronization primitives like `ZQueue` should be written in style that prevents data loss, expect some PRs :)

For `for` thing is indeed nasty, that's why I stopped using `for` altogether in favor of my macro library (https://github.com/mvv/sash).

Reply:
> A remaining question is whether `interruptible(uninterruptible(foo)) === uninterruptible(foo)`, and `uninterruptible(interruptible(foo)) === interruptible(foo)`.

I think the second one should hold, but the first one should not, it would be consistent with
`interruptible(/* IN HERE */ IO.succeed(x))` being interruptible.